RATEST-74: Fixing FormTests

sherrif10 opened the pull request:
https://issues.openmrs.org/browse/RATEST-74    cc @k-joseph  @kdaud

kdaud commented:
Refactor functions within `addFormTest()`, `editFormTest()`, `deleteFormTest()` that can be `re-usable` and create a separate function to rather invoke it in those respective methods to eliminate `code` redundancy.

sherrif10 commented:
Thanks @kdaud . i had actually refactored all those methods to only include sine method reference, However one thing we are catering here is the workflow, these tests are not synchronously running, so thats why included each mthod each test but with in one .Will still see if i can refactor more to look cleaner

kdaud commented:
> However one thing we are catering here is the workflow, these tests are not synchronously running, so thats why included each mthod each test but with in one

Technically i see the code block below being consumed by those three respective function

 {
`{
 homePage.goToManageForm();
        if (!manageForm.addPresent()) {
            manageForm.delete();
        }
        manageForm.add();
        manageForm.addLabel("Eye Report");
        manageForm.addIcon("icon-align-justify");
        manageForm.formIdFromUrl();
        manageForm.save();
        headerPage.clickOnHomeIcon();
        homePage.goToManageForm();
        manageForm.waitForPage()
}`

This is code [redundancy](https://dzone.com/articles/your-code-is-redundant-live-with-it) in programming. You rather refactor it as I mentioned in my previous comment.

sherrif10 commented:
Thats very true, Feel free to look at my recent commit, so here the code looks almost the same but the workflow changes,Forexample on editFormTest, However i think my last commit addresses your point, i have refactored addFormTest and InitiateFormPresence into one , Again one reason i added the implementation of addFormTest is because am catering where the server is clean without a new form,  since these tests run randonmly or asynchronoue

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/ActiveVisitsPage.java:
```diff
@@ -44,6 +45,7 @@ public void search(String text) {
     }
 
     public ClinicianFacingPatientDashboardPage goToPatientDashboardOfLastActiveVisit() {
+    	waiter.until(ExpectedConditions.visibilityOfElementLocated(ACTIVE_PATIENT));
```
Fix the indention here

sherrif10 replied:
There is no identation here, this is the same identation this project had before

kdaud replied:
You may need to cross the existing code indention otherwise `waiter` method is not well indented.

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/HeaderPage.java:
```diff
@@ -24,6 +25,7 @@ public void clickOnHomeLink() {
     }
 
     public void logOut() throws InterruptedException {
+    	waiter.until(ExpectedConditions.visibilityOfElementLocated(LOGOUT));
```
The same here

sherrif10 replied:
No identation to be applied here

kdaud replied:
See my comment [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/503/commits/20aa60c34874632e6109b53f2d9370bc6f5538d7#r675345648)

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/HtmlFormsPage.java:
```diff
@@ -0,0 +1,28 @@
+        super(page);
+    }
+
+    public void createNewFormTest(String name, String description, String version) throws InterruptedException {
```
Any reason for adding `InterruptedException` ?

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/page/ManageHtmlFormsPage.java:
```diff
@@ -0,0 +1,24 @@
+import org.openqa.selenium.support.ui.ExpectedConditions;
+
+public class ManageHtmlFormsPage extends Page {
+    private static final By NEW_HTML_FORM = By.cssSelector("a[href='htmlForm.form']");
```
Refer to the artifacts collected [here](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/476#issuecomment-870471982) to ensure quality assurance to align with software standards

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Created by nata on 24.06.15.
```
Git tracks the java class `[FormTest]` to be newly created and am wondering whether `nata` is the java class author !

kdaud replied:
Any response @sherrif10 on the comment above

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+	@Test
+	@Category(BuildTests.class)
+	public void addFormTest() throws Exception {
+		
```
Remove the redundant line` LN-64`

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+	@Test
+	@Category(BuildTests.class)
+	public void editFormTest() throws Exception {
+
```
Even `LN-78` should be removed

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+	@Test
+	@Category(BuildTests.class)
+	public void deleteFormTest() throws Exception {
+
```
`LN-92` should be removed too

sherrif10 commented:
cc @ibacher  @dkayiwa  Requesting for review thanks

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+		htmlFormsPage = new HtmlFormsPage(page);
+		manageHtmlFormsPage = new ManageHtmlFormsPage(page);
+		patientDashboardPage = new ClinicianFacingPatientDashboardPage(page);
+	}
```
You need to indent `setUp() `method to match with the existing ui test indention

sherrif10 replied:
Where exactly need to  be indented, these identation is the same as whole commit

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+		homePage.goToManageForm();
+		if (!manageForm.addPresent()) {
+			manageForm.delete();
+		}
```
Address the indention in the method as well.

kdaud commented:
Thanks @sherrif10 for the work. Catch my reviews above

kdaud commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,108 @@
+		manageHtmlFormsPage = new ManageHtmlFormsPage(page);
+		patientDashboardPage = new ClinicianFacingPatientDashboardPage(page);
+	}
+
```
There might be a typo in the method name below

sherrif10 commented:
cc @ibacher  @dkayiwa @kdaud

ibacher commented on ui-tests/src/test/java/org/openmrs/reference/FormTest.java:
```diff
@@ -0,0 +1,117 @@
+	public void deleteFormTest() throws Exception {
+		
+		addFormTest();
+		headerPage.clickOnHomeIcon();
```
Shouldn't we already have done this by calling `addFormTest()`?

sherrif10 replied:
We decided to merge all the three functionality, AddForm, editForm And DeleteForm simply because we needed to have test independent,because if we had separated them , each test would be dependant which would not be appropriate to handle while running each specific test

ibacher replied:
I get that and it makes sense. It's just that the last active line of `addFormTest()` is [`headerPage.clickOnHomeIcon()`](https://github.com/openmrs/openmrs-distro-referenceapplication/pull/503/files/da2b612b843e37731990ac581354c07e4a2a92f6#diff-40f58e09e158a43ef8df5dc6de3ea2f206cca3b6dc09552fd200ddea0ef795bbR80), so, in other words, isn't this test already starting with this having been done?

sherrif10 replied:
ohh yeah true , Let me refactor that

ibacher commented on ui-tests/src/test/java/org/openmrs/reference/page/HtmlFormsPage.java:
```diff
@@ -0,0 +1,28 @@
+import org.openmrs.uitestframework.page.Page;
+import org.openqa.selenium.By;
+
+public class HtmlFormsPage extends Page {
```
nit: since this is a new file, it should be indented with tabs, not spaces.

ibacher commented on ui-tests/src/test/java/org/openmrs/reference/page/ManageHtmlFormsPage.java:
```diff
@@ -0,0 +1,24 @@
+package org.openmrs.reference.page;
+
+import org.openmrs.uitestframework.page.Page;
```
nit: since this is a new file, it should be indented with tabs, not spaces.

sherrif10 replied:
Resolved thanks

ibacher reviewed (commented):
Apart from some minor issues, this looks fine to me.
